github.com/cockroachdb/pebble/internal/metamorphic: TestMetaTwoInstance failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/17234171?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/17234171?buildTab=artifacts#meta) on refs/heads/master @ [e1b330bd4290](https://github.com/cockroachdb/pebble/commits/e1b330bd4290):


```
        db1.Get("dbmhsl") // [""] pebble: not found #1208
        db1.DeleteRange("kpcq@2", "mdvu") // <nil> #1216
        db1.RangeKeyDelete("oxupyyw", "yfrs") // <nil> #1217
        db1.Get("olxbtmvpqi@1") // [""] pebble: not found #1218
        db1.Set("zunysqr", "i") // <nil> #1220
        db1.Delete("mtnxza@2") // <nil> #1226
        db1.LogData("uf") // <nil> #1227
        db1.Set("ghgc", "btzcfvowjhwky") // <nil> #1234
        iter10.Prev("") // [false] <nil> #1036
        iter8.First() // [true,"qvatnvjvple",<no point>,["qvatnvjvple","uocvzlcbo")=>{""="xpzaqh"}*] <nil> #973
        iter8.Last() // [true,"yfrs",<no point>,["yfrs","yvnaf")=>{"@2"="jdgmebgus","@1"="olwbjiouoilqq"}*] <nil> #974
        iter10.SeekPrefixGE("tbdism@2") // [true,"tbdism@2",<no point>,["tbdism","tbdism\x00")=>{""="xpzaqh"}*] <nil> #1041
        iter10.Prev("") // [false] pebble: unsupported reverse prefix iteration #1044
        iter10.Prev("zunysqr@2") // [invalid] pebble: unsupported reverse prefix iteration #1046
        iter10.Next("") // [false] pebble: unsupported reverse prefix iteration #1048
        iter8.SeekGE("sujhk@1", "") // [true,"sujhk@1",<no point>,["qvatnvjvple","uocvzlcbo")=>{""="xpzaqh"}*] <nil> #975
        iter10.SeekGE("yvxpiuoljfn@1", "") // [false] <nil> #1049
        iter8.SeekGE("pxfhl@2", "") // [true,"qvatnvjvple",<no point>,["qvatnvjvple","uocvzlcbo")=>{""="xpzaqh"}] <nil> #976
        iter8.Prev("") // [false] <nil> #977
        iter10.SeekPrefixGE("tbdism@2") // [true,"tbdism@2",<no point>,["tbdism","tbdism\x00")=>{""="xpzaqh"}*] <nil> #1053
        iter10.SeekPrefixGE("rfaembhcnls@2") // [true,"rfaembhcnls@2","rapwtbrpbnequacujem",["rfaembhcnls@2","rfaembhcnls\x00")=>{""="xpzaqh"}*] <nil> #1057
        iter10.First() // [true,"rfaembhcnls@2","rapwtbrpbnequacujem",["rfaembhcnls@2","uocvzlcbo")=>{""="xpzaqh"}*] <nil> #1059
        iter10.SeekPrefixGE("tbdism@2") // [true,"tbdism@2",<no point>,["tbdism","tbdism\x00")=>{""="xpzaqh"}*] <nil> #1063
        iter10.NextPrefix() // [false] NextPrefix not permitted with upper bound zunysqr@2 #1065
        iter10.Prev("") // [false] NextPrefix not permitted with upper bound zunysqr@2 #1066
        iter10.Next("teyxgqcg@2") // [invalid] cannot use limit with prefix iteration #1071
        iter10.NextPrefix() // [false] NextPrefix not permitted with upper bound zunysqr@2 #1078
        iter10.SeekLT("dkps@1", "") // [false] <nil> #1081
        iter10.SeekLT("ghgc", "") // [false] <nil> #1083
        iter10.SeekPrefixGE("rfaembhcnls@2") // [true,"rfaembhcnls@2","rapwtbrpbnequacujem",["rfaembhcnls@2","rfaembhcnls\x00")=>{""="xpzaqh"}*] <nil> #1085
        iter10.SeekPrefixGE("rfaembhcnls@2") // [true,"rfaembhcnls@2","rapwtbrpbnequacujem",["rfaembhcnls@2","rfaembhcnls\x00")=>{""="xpzaqh"}] <nil> #1086
        iter10.Prev("") // [false] pebble: unsupported reverse prefix iteration #1087
        iter10.SeekGE("oxupyyw@2", "") // [true,"rfaembhcnls@2","rapwtbrpbnequacujem",["rfaembhcnls@2","uocvzlcbo")=>{""="xpzaqh"}*] <nil> #1092
        iter10.SeekLT("wwjwkgmui@1", "") // [true,"veoezrkazqt",<no point>,["veoezrkazqt","yfrs")=>{"@2"="jdgmebgus"}*] <nil> #1093
        iter10.First() // [true,"rfaembhcnls@2","rapwtbrpbnequacujem",["rfaembhcnls@2","uocvzlcbo")=>{""="xpzaqh"}*] <nil> #1094
        iter10.SeekLT("qvatnvjvple@1", "") // [false] <nil> #1097
        iter10.Last() // [true,"yfrs",<no point>,["yfrs","yvnaf")=>{"@1"="olwbjiouoilqq"}*] <nil> #1098
        iter10.First() // [true,"rfaembhcnls@2","rapwtbrpbnequacujem",["rfaembhcnls@2","uocvzlcbo")=>{""="xpzaqh"}*] <nil> #1103
        iter10.SeekLT("eydhihouc@1", "") // [false] <nil> #1108
        iter10.SeekGE("lioknhay@2", "") // [true,"rfaembhcnls@2","rapwtbrpbnequacujem",["rfaembhcnls@2","uocvzlcbo")=>{""="xpzaqh"}*] <nil> #1114
        iter10.Prev("") // [false] <nil> #1117
        iter10.Last() // [true,"yfrs",<no point>,["yfrs","yvnaf")=>{"@1"="olwbjiouoilqq"}*] <nil> #1121
        iter10.First() // [true,"rfaembhcnls@2","rapwtbrpbnequacujem",["rfaembhcnls@2","uocvzlcbo")=>{""="xpzaqh"}*] <nil> #1123
        iter11.Last() // [true,"yfrs",<no point>,["yfrs","yvnaf")=>{"@1"="olwbjiouoilqq"}*] <nil> #1119
        iter11.First() // [true,"bbeuzkdm",<no point>,["bbeuzkdm","bzrzysxvhayh")=>{"@2"="ccbmjjprgufvxsxarg"}*] <nil> #1120
        iter10.SetBounds("zunysqr@2", "zunysqr@2") // <nil> #1126
        iter10.SeekGE("zunysqr@2", "") // [false] <nil> #1127
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/241009-041227.9611872548339/standard-020 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/standard-020 (0.10s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1728447147961814229 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: PEBBLE-271

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/17252314?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/17252314?buildTab=artifacts#meta) on refs/heads/master @ [27acbe1d595c](https://github.com/cockroachdb/pebble/commits/27acbe1d595c):


```
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000072
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000073
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000074
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000075
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000077
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000078
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000079
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000080
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000081
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000085
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000093
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000097
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000103
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000114
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000115
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000116
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000117
        // 04:10:20.359 INFO: [JOB 118] sstable deleted 000118
        // 04:10:20.360 INFO: [JOB 118] sstable deleted 000119
        // 04:10:20.360 INFO: [JOB 118] sstable deleted 000120
        // 04:10:20.360 INFO: [JOB 118] sstable deleted 000121
        // 04:10:20.360 INFO: [JOB 118] MANIFEST deleted 000133
        // 04:10:20.360 INFO: [JOB 119] sstable deleted 000106
        // 04:10:20.360 INFO: [JOB 119] sstable deleted 000153
        // 04:10:20.360 INFO: [JOB 120] sstable deleted 000107
        // 04:10:20.360 INFO: [JOB 120] sstable deleted 000154
        // 04:10:20.360 INFO: [JOB 124] MANIFEST deleted 000134
        // 04:10:20.360 INFO: [JOB 132] sstable deleted 000108
        // 04:10:20.360 INFO: [JOB 132] sstable deleted 000111
        // 04:10:20.360 INFO: [JOB 132] sstable deleted 000155
        // 04:10:20.360 INFO: [JOB 132] sstable deleted 000159
        // 04:10:20.360 INFO: [JOB 131] sstable deleted 000112
        // 04:10:20.360 INFO: [JOB 131] sstable deleted 000160
        // 04:10:20.360 INFO: [JOB 135] sstable deleted 000113
        // 04:10:20.360 INFO: [JOB 135] sstable deleted 000161
        // 04:10:20.360 INFO: [JOB 136] sstable deleted 000104
        // 04:10:20.360 INFO: [JOB 136] sstable deleted 000151
        // 04:10:20.360 INFO: [JOB 139] sstable deleted 000110
        // 04:10:20.360 INFO: [JOB 139] sstable deleted 000158
        // 04:10:20.363 INFO: [JOB 141] compacted(default) L0 [000102 000149] (1.8KB) Score=125.00 + L1 [] (0B) Score=0.00 -> L1 [000174] (923B), in 0.0s (0.0s total), output rate 2.9MB/s
        // 04:10:20.363 INFO: [JOB 143] compacting(move) multilevel L1 [000174] (923B) Score=11537.50 + L2 [] (0B) Score=0.00 + L3 [] (0B) Score=3.97; OverlappingRatio: Single 0.00, Multi 0.00
        // 04:10:20.363 INFO: [JOB 143] compacted(move) multilevel L1 [000174] (923B) Score=11537.50 + L2 [] (0B) Score=0.00 + L3 [] (0B) Score=3.97 -> L3 [000174] (923B), in 0.0s (0.0s total), output rate 442MB/s
        // 04:10:20.363 INFO: [JOB 144] compacting(default) L0 [000105 000152] (1.8KB) Score=112.50 + L1 [] (0B) Score=0.00; OverlappingRatio: Single 0.00, Multi 0.00
        // 04:10:20.364 INFO: [JOB 144] compacting: sstable created 000175
        // 04:10:20.364 INFO: [JOB 141] sstable deleted 000102
        // 04:10:20.364 INFO: [JOB 141] sstable deleted 000149
        // 04:10:20.366 INFO: [JOB 116] MANIFEST created 000180
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/241010-041019.3361854394044/random-025 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-025 (0.36s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1728533419336389228 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/compaction.go
+++ b/compaction.go
@@ -2562,15 +2562,6 @@ func (d *DB) runDeleteOnlyCompactionForLevel(
 			panic("pebble: delete-only compaction scheduled with hints that did not delete or excise a file")
 		}
 	}
-	// Remove any files that were added and deleted in the same versionEdit.
-	ve.NewFiles = slices.DeleteFunc(ve.NewFiles, func(e manifest.NewFileEntry) bool {
-		deletedFileEntry := manifest.DeletedFileEntry{Level: cl.level, FileNum: e.Meta.FileNum}
-		if _, deleted := ve.DeletedFiles[deletedFileEntry]; deleted {
-			delete(ve.DeletedFiles, deletedFileEntry)
-			return true
-		}
-		return false
-	})
 	return nil
 }
 
@@ -2633,6 +2624,27 @@ func (d *DB) runDeleteOnlyCompaction(
 		}
 		c.metrics[cl.level] = levelMetrics
 	}
+	// Remove any files that were added and deleted in the same versionEdit.
+	ve.NewFiles = slices.DeleteFunc(ve.NewFiles, func(e manifest.NewFileEntry) bool {
+		deletedFileEntry := manifest.DeletedFileEntry{Level: e.Level, FileNum: e.Meta.FileNum}
+		if _, deleted := ve.DeletedFiles[deletedFileEntry]; deleted {
+			delete(ve.DeletedFiles, deletedFileEntry)
+			return true
+		}
+		return false
+	})
+	// Remove any entries from CreatedBackingTables that are not used in any
+	// NewFiles.
+	usedBackingFiles := make(map[base.DiskFileNum]struct{})
+	for _, e := range ve.NewFiles {
+		if e.Meta.Virtual {
+			usedBackingFiles[e.Meta.FileBacking.DiskFileNum] = struct{}{}
+		}
+	}
+	ve.CreatedBackingTables = slices.DeleteFunc(ve.CreatedBackingTables, func(b *fileBacking) bool {
+		_, used := usedBackingFiles[b.DiskFileNum]
+		return !used
+	})
 	// Refresh the disk available statistic whenever a compaction/flush
 	// completes, before re-acquiring the mutex.
 	d.calculateDiskAvailableBytes()
@@ -2683,6 +2695,15 @@ func (d *DB) runCompaction(
 	}
 	switch c.kind {
 	case compactionKindDeleteOnly:
+		// Before dropping the db mutex, grab a ref to the current version. This
+		// prevents any concurrent excises from deleting files that this compaction
+		// needs to read/maintain a reference to.
+		//
+		// Note that delete-only compactions can call excise(), which needs to be able
+		// to read these files.
+		vers := d.mu.versions.currentVersion()
+		vers.Ref()
+		defer vers.UnrefLocked()
 		// Release the d.mu lock while doing I/O.
 		// Note the unusual order: Unlock and then Lock.
 		snapshots := d.mu.snapshots.toSlice()

